# Data Cleaning

This notebook cleans the dataset by removing NaN values and filtering out URLs and other irrelevant content.

In [1]:
import pandas as pd

input_file_path = '/Users/thebekhruz/Desktop/100Days-Of-Code/100-Days-of-NLP-Odyssey/data/raw/swop_triples.csv'
output_file_path = '/Users/thebekhruz/Desktop/100Days-Of-Code/100-Days-of-NLP-Odyssey/data/intermediate/swop_triples_cleaned.csv'


## Initial Data Exploration:

In [2]:
df = pd.read_csv(input_file_path, delimiter='\t', header=None, names=['doc_id', 'type', 'value'])
print(df.shape)

(169958, 3)


After exploration there only should be 3 unique values

In [3]:
#
print(df['type'].unique())
 

['title' 'description' 'mentions' 'tanc:became' 'tanc:showing'
 'tanc:gardens' 'tanc:surrounded' 'tanc:built' 'tanc:viewed' nan
 'tanc:displayed' 'tanc:inviting' 'tanc:used_to_carry' 'tanc:standing'
 'tanc:showing_shelving' 'tanc:setting' 'tanc:opened' 'tanc:known'
 'tanc:presented' 'tanc:signing' 'tanc:speaking' 'tanc:watch'
 'tanc:helping' 'tanc:visiting' 'tanc:including' 'tanc:sponsored'
 'tanc:fought' 'tanc:stands' 'tanc:preparing_to_take' 'tanc:seen'
 'tanc:gives' 'tanc:came_to_give' 'tanc:also' 'tanc:watched' 'tanc:set'
 'tanc:sold' 'tanc:shows' 'tanc:succeeded' 'tanc:flanked' 'tanc:pulling'
 'tanc:got' 'tanc:having' 'tanc:posing' 'tanc:operating' 'tanc:lists'
 'tanc:put' 'tanc:enter' 'tanc:sitting' 'tanc:in_comparison'
 'tanc:brides_to_leave' 'tanc:bought' 'tanc:giving' 'tanc:joined'
 'tanc:handing' 'tanc:moved' 'tanc:connects' 'tanc:give' 'tanc:putting'
 'tanc:dates' 'tanc:restored' 'tanc:situated' 'tanc:entertaining'
 'tanc:holding' 'tanc:starring' 'tanc:just' 'tanc:parked' 't

169,958 entries and 3 types of data. 

In [4]:
df.isna().value_counts()

doc_id  type   value
False   False  False    168896
               True        574
        True   True        488
Name: count, dtype: int64

- 168,896, records have no missing values across all three columns (doc_id, type, value).
- 574 records where doc_id and type are present, but value is missing (NaN)
- 488 records are missing both type and value, with only the doc_id being present.

In [5]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169958 entries, 0 to 169957
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   doc_id  169958 non-null  object
 1   type    169470 non-null  object
 2   value   168896 non-null  object
dtypes: object(3)
memory usage: 3.9+ MB
None


## Filtering rows with NaN

In [6]:
# Explore rows with NaN values
rows_with_nan = df[df.isna().any(axis=1)]
rows_with_nan.head(10)

doc_id         type value
1187  This view shows the third and final phase of t...          NaN   NaN
1726  The foundation stone was laid by the Chairman ...          NaN   NaN
2446                The same event is shown in BFP01772          NaN   NaN
3079  This is likely to be The Crown at Wheeler End ...          NaN   NaN
3282  The group was started by the Holmer Green Wome...          NaN   NaN
3646               59f99c46-3663-3740-ac46-2e9479a773ab  description   NaN
3683  The man and woman on the outside are Donald an...          NaN   NaN
3684  Accompanying them are Gwen and Sidney Hall. He...          NaN   NaN
3685  They were all at a party given by Sir Francis ...          NaN   NaN
3970  Although the event took place in West Wycombe,...          NaN   NaN

In [7]:
# in-depth analysis of NaN values
for column in df.columns:
    print(f"Rows with NaN in {column}:")
    print(df[df[column].isna()])


Rows with NaN in doc_id:
Empty DataFrame
Columns: [doc_id, type, value]
Index: []
Rows with NaN in type:
                                                   doc_id type value
1187    This view shows the third and final phase of t...  NaN   NaN
1726    The foundation stone was laid by the Chairman ...  NaN   NaN
2446                  The same event is shown in BFP01772  NaN   NaN
3079    This is likely to be The Crown at Wheeler End ...  NaN   NaN
3282    The group was started by the Holmer Green Wome...  NaN   NaN
...                                                   ...  ...   ...
152444  The four children in the bottom right of the i...  NaN   NaN
157305  This is one of a series of photos (BFP78468 to...  NaN   NaN
159312                                  See also RHW31136  NaN   NaN
159542                                 See also RHW31138.  NaN   NaN
159586                                 See also RHW31138.  NaN   NaN

[488 rows x 3 columns]
Rows with NaN in value:
                   

- The doc_id column has no rows with missing values.
- The type column has 488 rows with missing values, where both type and value columns are NaN. These entries also include textual content in the doc_id column, suggesting a possible misalignment or data entry issue.
- The value column has 1,062 rows with missing values. This includes the 488 rows where both type and value are missing, and an additional 574 rows where only value is missing, but doc_id and type are present.

# Identify and Handle Missing Values

#### Remove 488 rows where both `type` and `value` columns are NaN

In [8]:
# An example of a row where 'type' is NaN
example_row_with_nan_type = df[df['type'].isna()].head(5)
print(example_row_with_nan_type)

                                                 doc_id type value
1187  This view shows the third and final phase of t...  NaN   NaN
1726  The foundation stone was laid by the Chairman ...  NaN   NaN
2446                The same event is shown in BFP01772  NaN   NaN
3079  This is likely to be The Crown at Wheeler End ...  NaN   NaN
3282  The group was started by the Holmer Green Wome...  NaN   NaN


> I am removing all rows where the doc_id is clearly a description that has no relation to any actual doc_id.

In [9]:
# Remove rows where `type` and `value` is NaN
df.dropna(subset=['type'], inplace=True)
df.head()

doc_id         type  \
0  a7bb9917-95ff-3f55-a640-4c5afcec25f2        title   
1  a7bb9917-95ff-3f55-a640-4c5afcec25f2  description   
2  a7bb9917-95ff-3f55-a640-4c5afcec25f2     mentions   
3  a7bb9917-95ff-3f55-a640-4c5afcec25f2     mentions   
4  c29a7b77-7c46-3b85-88fe-05c8f4b2e384        title   

                                               value  
0  View towards SE of junction of Queen Victoria ...  
1  E corner of Queen Victoria Rd at junction with...  
2               https://www.wikidata.org/wiki/Q64116  
3               https://www.wikidata.org/wiki/Q64116  
4  Front page of Bucks Free Press, Time capsule f...

In [10]:
# An example of a row where 'value' is NaN
example_row_with_nan_value = df[df['value'].isna()].head(5)
print(example_row_with_nan_value)

                                    doc_id         type value
3646  59f99c46-3663-3740-ac46-2e9479a773ab  description   NaN
8765  9ea24c7c-599f-32f7-a17e-cf2ad800fdf0  description   NaN
8792  64e24ebb-a966-3298-a2f3-6684a0346f7a  description   NaN
9141  3ac5ab74-d10e-37c6-ad2e-a9da9f33b77c  description   NaN
9178  971934e5-a771-3739-a1ba-f1b4827874cf  description   NaN


> Sometimes the items in the dataset only contain titles and no description. We need to remove them.

In [11]:
# Replace NaN in description with an empty string
df.fillna({'value': ''}, inplace=True)

In [12]:
print(df.isna().sum())


doc_id    0
type      0
value     0
dtype: int64


> The output indicates there are no missing values (NaN) across all columns (doc_id, type, value) in the DataFrame.

# Validate 'doc_id' Format

In [13]:
uuid_regex = r'^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$'

# Use str.match to check each doc_id against the UUID regex
matches_uuid_format = df['doc_id'].str.match(uuid_regex)
not_matching = df[~matches_uuid_format]
print(not_matching)


# To get a count of how many doc_id values do not match the UUID format
not_matching_count = not_matching.shape[0]
print(f"Count of doc_id values not matching UUID format: {not_matching_count}")

                                         doc_id              type  \
29         https://www.wikidata.org/wiki/Q64116       tanc:became   
244        https://www.wikidata.org/wiki/Q19695      tanc:showing   
357     https://www.wikidata.org/wiki/Q12823105      tanc:gardens   
382      https://www.wikidata.org/wiki/Q8039639      tanc:showing   
561        https://www.wikidata.org/wiki/Q64116   tanc:surrounded   
...                                         ...               ...   
169200   https://www.wikidata.org/wiki/Q1541030         tanc:gain   
169421       https://www.wikidata.org/wiki/Q222  tanc:transferred   
169662   https://www.wikidata.org/wiki/Q5378261      tanc:dancing   
169776   https://www.wikidata.org/wiki/Q8039639      tanc:singing   
169954   https://www.wikidata.org/wiki/Q5378261      tanc:dancing   

                                         value  
29      https://www.wikidata.org/wiki/Q8034980  
244       https://www.wikidata.org/wiki/Q64116  
357     https://www.wiki

> There are 993 doc_id values in the DataFrame that do not match the UUID format specified by the regex pattern. 

In [14]:
# Remove fields in doc_id which are not in the form of UUID
df = df[matches_uuid_format]

# Check the updated DataFrame
print(df.shape)


(168477, 3)


In [15]:
print(df['type'].unique())


['title' 'description' 'mentions']


As mentioned above only 3 unique values and therefore clean data. 

### Export Cleaned Data

In [16]:
df.to_csv(output_file_path, sep='\t', index=False, header=None)
